### ISSUE TITLE
Consider using a CDN that is resistant to domain fronting

### ISSUE THREAD

**Issue by rhuddleston** [pypi/warehouse#10399]
**What's the problem this feature will solve?**

https://arstechnica.com/information-technology/2021/11/malware-downloaded-from-pypi-41000-times-was-surprisingly-stealthy/

Pypi currently uses fastly CDN which apparently has no safeguards against "domain fronting" this means I can register a fastly domain and use it as my command and control server for malware but use the pypi.org domain to funnel this traffic. This mean even though I have a restricted firewall that prevents all but whitelisted domains e.g. pypi.org is in the whitelist, I can not prevent malware from getting downloaded and controlled through the pypi.org domain.

example of domain fronting: 

NY Times image downloaded via fastly: 

https://static01.nyt.com/images/2021/11/21/multimedia/21xp-Cold-case/21xp-Cold-case-threeByTwoSmallAt2X.jpg

The same image downloaded via pypi.org: 

curl -H 'Host: static01.nyt.com' https://pypi.org/images/2021/11/21/multimedia/21xp-Cold-case/21xp-Cold-case-threeByTwoSmallAt2X.jpg --output 21xp-Cold-case-threeByTwoSmallAt2X.jpg

**Describe the solution you'd like**

CDN providers such as AWS Cloudfront and Cloudflare (Google Cloud as well is my understanding) prevent domain fronting consider using one of these providers instead or another one that prevents this. As an alternative see if fastly has any intention of enabling this or at least giving an option if you want to enforce the Host header and SNI to match for your domain.

Thanks for considering this to improve Pypi's security posture

**Comment by abitrolly** [comment]
Wow. It looks like a misconfiguration on the PyPI side. However, I believe that the configuration is not present in this repo, so it is hard to validate it.

And I also remember that Fastly provided the service for free, so using an alternative may be too expensive. Need someone to confirm that,

**Comment by rhuddleston** [comment]
cloudflare has a free version as well FWIW but if you can configure fastly to enforce SNI to match hostname that should also work.

**Comment by ewjoachim** [comment]
Wether we like it or not, PyPi is in a pretty unique position on the internet as a critical infrastructure service that's not backed by a company and with limited funding. I think that before considering switching providers, a good first approach is to speak to people. Fastly is an infrastructure sponsor. We're not just using their free tier, we're using their full service and they make it free for us because it's us.

So I'd say first: have we spoken to fastly about this ? Is it something they can/accept to/will solve ? If not, and if we consider it's a bigger risk to leave this door open, then we can evaluate other providers, but I'm pretty sure that given our bandwidth requirements, we're not "just" going to jump to another service's free tier.

**Comment by abitrolly** [comment]
@ewjoachim I feel there is something wrong about the whole situation that for a freely provided community service, we don't know people in community who championed the service to speak about it.

**Comment by abitrolly** [comment]
> I'm pretty sure that given our bandwidth requirements, we're not "just" going to jump to another service's free tier.

For the bandwidth requirement I still hope to reopen https://github.com/pypa/warehouse/issues/10355 as its dupe predecessor is not closed in 6 years despite having high priority label.

**Comment by ewjoachim** [comment]
I don't think we need live traffic data integrated in the tool in order to have an idea of how much we consume.

See https://dustingram.com/articles/2021/04/14/powering-the-python-package-index-in-2021/ for figures :)

> Fastly gives us a 100% discount on our bill, which at the time of writing would be more than $1.8M (yes, million) per month.

**Comment by abitrolly** [comment]
@ewjoachim the rationale why I need the traffic data is outlined in #10355. Just because sponsors provide resources, doesn't mean PyPI should not preserve these. As a sponsor I would expect that PyPI not only preserve resources, but also document how to do this in a proper way to be an exemplary project with best practices for others in Python ecosystem, who can not enjoy $300000 maintenance funding in addition to free resources.

**Comment by dstufft** [comment]
I think we could implement this ourselves in VCL with

```vcl
if tls.client.servername !=req.http.host {
    error XXX "..."
}
```

**Comment by miketheman** [comment]
Fastly changed to disallow domain fronting in 2024. One email they sent out to customers: https://archive.torproject.org/websites/lists.torproject.org/pipermail/anti-censorship-team/2023-October/000328.html

The same curl request today:

```bash
$ curl -H 'Host: static01.nyt.com' https://pypi.org/images/2021/11/21/multimedia/21xp-Cold-case/21xp-Cold-case-threeByTwoSmallAt2X.jpg
Requested host does not match any Subject Alternative Names (SANs) on TLS certificate [2ffb92323fe37bc835f67c361f86162dfbad788d8ec039220b0c512646ee86fb] in use with this connection.

Visit https://docs.fastly.com/en/guides/common-400-errors#error-421-misdirected-request for more information.
```

Here's their docs section on the error:
https://www.fastly.com/documentation/guides/concepts/errors/#routing-errors

With this protection in place, closing! If you have further evidence of domain fronting, please open a fresh issue with details (and feel free to reference this issue if relevant).
